File: backend/models/Notification.js
const pool = require('../config/database');

class Notification {
  // Get all notifications for a user
  static async findByUserId(userId, options = {}) {
    const { isRead, type, limit = 50, offset = 0 } = options;
    
    let query = 'SELECT * FROM notifications WHERE user_id = ?';
    const params = [userId];
    
    if (isRead !== undefined) {
      query += ' AND is_read = ?';
      params.push(isRead);
    }
    
    if (type) {
      query += ' AND type = ?';
      params.push(type);
    }
    
    query += ' ORDER BY created_at DESC LIMIT ? OFFSET ?';
    params.push(limit, offset);
    
    const [rows] = await pool.query(query, params);
    return rows;
  }

  // Get unread count for a user
  static async getUnreadCount(userId) {
    const [rows] = await pool.query(
      'SELECT COUNT(*) as count FROM notifications WHERE user_id = ? AND is_read = FALSE',
      [userId]
    );
    return rows[0].count;
  }

  // Create a notification
  static async create(notificationData) {
    const { userId, title, message, type = 'system', data = null } = notificationData;
    
    if (!userId || !title || !message) {
      throw new Error('Missing required fields: userId, title, message');
    }

    const validTypes = ['order', 'promotion', 'system', 'review', 'account'];
    if (!validTypes.includes(type)) {
      throw new Error(`Invalid notification type. Must be one of: ${validTypes.join(', ')}`);
    }

    const [result] = await pool.query(
      `INSERT INTO notifications (user_id, title, message, type, data) 
       VALUES (?, ?, ?, ?, ?)`,
      [userId, title, message, type, data ? JSON.stringify(data) : null]
    );
    
    return this.findById(result.insertId);
  }

  // Create multiple notifications (bulk insert)
  static async createBulk(userIds, title, message, type = 'system', data = null) {
    if (!Array.isArray(userIds) || userIds.length === 0) {
      throw new Error('userIds must be a non-empty array');
    }

    if (!title || !message) {
      throw new Error('Missing required fields: title, message');
    }

    const validTypes = ['order', 'promotion', 'system', 'review', 'account'];
    if (!validTypes.includes(type)) {
      throw new Error(`Invalid notification type. Must be one of: ${validTypes.join(', ')}`);
    }

    const values = userIds.map(userId => [
      userId,
      title,
      message,
      type,
      data ? JSON.stringify(data) : null
    ]);

    const [result] = await pool.query(
      `INSERT INTO notifications (user_id, title, message, type, data) 
       VALUES ?`,
      [values]
    );

    return {
      success: true,
      insertedCount: result.affectedRows,
      insertId: result.insertId
    };
  }

  // Find notification by ID
  static async findById(id) {
    const [rows] = await pool.query(
      'SELECT * FROM notifications WHERE id = ?',
      [id]
    );
    return rows[0];
  }

  // Mark notification as read
  static async markAsRead(id, userId) {
    const [result] = await pool.query(
      'UPDATE notifications SET is_read = TRUE WHERE id = ? AND user_id = ?',
      [id, userId]
    );
    return result.affectedRows > 0;
  }

  // Mark all notifications as read for a user
  static async markAllAsRead(userId) {
    const [result] = await pool.query(
      'UPDATE notifications SET is_read = TRUE WHERE user_id = ? AND is_read = FALSE',
      [userId]
    );
    return result.affectedRows;
  }

  // Delete a notification
  static async delete(id, userId) {
    const [result] = await pool.query(
      'DELETE FROM notifications WHERE id = ? AND user_id = ?',
      [id, userId]
    );
    return result.affectedRows > 0;
  }

  // Delete all read notifications for a user
  static async deleteAllRead(userId) {
    const [result] = await pool.query(
      'DELETE FROM notifications WHERE user_id = ? AND is_read = TRUE',
      [userId]
    );
    return result.affectedRows;
  }

  // Get notification statistics for admin
  static async getStatistics() {
    const [stats] = await pool.query(`
      SELECT 
        COUNT(*) as total,
        SUM(CASE WHEN is_read = FALSE THEN 1 ELSE 0 END) as unread,
        SUM(CASE WHEN is_read = TRUE THEN 1 ELSE 0 END) as read,
        SUM(CASE WHEN type = 'order' THEN 1 ELSE 0 END) as order_notifications,
        SUM(CASE WHEN type = 'promotion' THEN 1 ELSE 0 END) as promotion_notifications,
        SUM(CASE WHEN type = 'system' THEN 1 ELSE 0 END) as system_notifications,
        SUM(CASE WHEN type = 'review' THEN 1 ELSE 0 END) as review_notifications,
        SUM(CASE WHEN type = 'account' THEN 1 ELSE 0 END) as account_notifications,
        COUNT(DISTINCT user_id) as unique_users,
        DATE(created_at) as date
      FROM notifications
      WHERE created_at >= DATE_SUB(NOW(), INTERVAL 30 DAY)
      GROUP BY DATE(created_at)
      ORDER BY date DESC
    `);
    
    return stats;
  }

  // Clean old notifications (older than 90 days)
  static async cleanOldNotifications(daysOld = 90) {
    const [result] = await pool.query(
      'DELETE FROM notifications WHERE is_read = TRUE AND created_at < DATE_SUB(NOW(), INTERVAL ? DAY)',
      [daysOld]
    );
    return result.affectedRows;
  }
}

module.exports = Notification;

File: frontend/lib/models/notification.dart
class AppNotification {
  final int id;
  final int userId;
  final String title;
  final String message;
  final NotificationType type;
  final bool isRead;
  final Map<String, dynamic>? data;
  final DateTime createdAt;
  final DateTime updatedAt;

  AppNotification({
    required this.id,
    required this.userId,
    required this.title,
    required this.message,
    required this.type,
    required this.isRead,
    this.data,
    required this.createdAt,
    required this.updatedAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'],
      userId: json['user_id'],
      title: json['title'],
      message: json['message'],
      type: NotificationType.fromString(json['type']),
      isRead: json['is_read'] == 1 || json['is_read'] == true,
      data: json['data'] != null
          ? (json['data'] is String
                ? {} // If data is string, use empty map (or parse JSON)
                : Map<String, dynamic>.from(json['data']))
          : null,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'title': title,
      'message': message,
      'type': type.value,
      'is_read': isRead,
      'data': data,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  AppNotification copyWith({
    int? id,
    int? userId,
    String? title,
    String? message,
    NotificationType? type,
    bool? isRead,
    Map<String, dynamic>? data,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return AppNotification(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      title: title ?? this.title,
      message: message ?? this.message,
      type: type ?? this.type,
      isRead: isRead ?? this.isRead,
      data: data ?? this.data,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  String getTimeAgo() {
    final now = DateTime.now();
    final difference = now.difference(createdAt);

    if (difference.inDays > 365) {
      final years = (difference.inDays / 365).floor();
      return '$years năm trước';
    } else if (difference.inDays > 30) {
      final months = (difference.inDays / 30).floor();
      return '$months tháng trước';
    } else if (difference.inDays > 0) {
      return '${difference.inDays} ngày trước';
    } else if (difference.inHours > 0) {
      return '${difference.inHours} giờ trước';
    } else if (difference.inMinutes > 0) {
      return '${difference.inMinutes} phút trước';
    } else {
      return 'Vừa xong';
    }
  }
}

enum NotificationType {
  order('order', 'Đơn hàng', '📦'),
  promotion('promotion', 'Khuyến mãi', '🎁'),
  system('system', 'Hệ thống', '🔔'),
  review('review', 'Đánh giá', '⭐'),
  account('account', 'Tài khoản', '👤');

  final String value;
  final String displayName;
  final String icon;

  const NotificationType(this.value, this.displayName, this.icon);

  static NotificationType fromString(String value) {
    return NotificationType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => NotificationType.system,
    );
  }
}

class NotificationListResponse {
  final List<AppNotification> notifications;
  final int unreadCount;
  final NotificationPagination pagination;

  NotificationListResponse({
    required this.notifications,
    required this.unreadCount,
    required this.pagination,
  });

  factory NotificationListResponse.fromJson(Map<String, dynamic> json) {
    return NotificationListResponse(
      notifications: (json['notifications'] as List)
          .map((item) => AppNotification.fromJson(item))
          .toList(),
      unreadCount: json['unreadCount'] ?? 0,
      pagination: NotificationPagination.fromJson(json['pagination']),
    );
  }
}

class NotificationPagination {
  final int limit;
  final int offset;
  final bool hasMore;

  NotificationPagination({
    required this.limit,
    required this.offset,
    required this.hasMore,
  });

  factory NotificationPagination.fromJson(Map<String, dynamic> json) {
    return NotificationPagination(
      limit: json['limit'] ?? 50,
      offset: json['offset'] ?? 0,
      hasMore: json['hasMore'] ?? false,
    );
  }
}

class NotificationStatistics {
  final int total;
  final int unread;
  final int read;
  final int orderNotifications;
  final int promotionNotifications;
  final int systemNotifications;
  final int reviewNotifications;
  final int accountNotifications;
  final int uniqueUsers;
  final DateTime date;

  NotificationStatistics({
    required this.total,
    required this.unread,
    required this.read,
    required this.orderNotifications,
    required this.promotionNotifications,
    required this.systemNotifications,
    required this.reviewNotifications,
    required this.accountNotifications,
    required this.uniqueUsers,
    required this.date,
  });

  factory NotificationStatistics.fromJson(Map<String, dynamic> json) {
    return NotificationStatistics(
      total: json['total'] ?? 0,
      unread: json['unread'] ?? 0,
      read: json['read'] ?? 0,
      orderNotifications: json['order_notifications'] ?? 0,
      promotionNotifications: json['promotion_notifications'] ?? 0,
      systemNotifications: json['system_notifications'] ?? 0,
      reviewNotifications: json['review_notifications'] ?? 0,
      accountNotifications: json['account_notifications'] ?? 0,
      uniqueUsers: json['unique_users'] ?? 0,
      date: DateTime.parse(json['date']),
    );
  }
}

class SendNotificationRequest {
  final List<int> userIds;
  final String title;
  final String message;
  final NotificationType type;
  final bool sendEmail;
  final Map<String, dynamic>? promotionData;

  SendNotificationRequest({
    required this.userIds,
    required this.title,
    required this.message,
    this.type = NotificationType.system,
    this.sendEmail = false,
    this.promotionData,
  });

  Map<String, dynamic> toJson() {
    return {
      'userIds': userIds,
      'title': title,
      'message': message,
      'type': type.value,
      'sendEmail': sendEmail,
      if (promotionData != null) 'promotionData': promotionData,
    };
  }
}
